### PR TITLE
Build fixes (-dotnet) for app extensions

### LINKED ISSUE #14285: iOS network extension crash on launch, `xamarin_extension_main` not initialized
### Steps to Reproduce

The repro I attached to #13742 suffices here as well, even though the symptom and apparent cause are different.

In this case, the iOS console log will show an entry from ReportCrash indicating an attempt to dereference a null pointer.

Since network extensions do not run in the simulator, the repro can only be shown on an actual iOS device.

Rather than reproducing the crash and seeing the message in the console log, we could think of the "repro" another way:

I *think* the problem is happening in `xamarin_main()` in `runtime/monotouch-main.m` at the call to `xamarin_extension_main`, which is actually a function pointer that is supposed to be initialized in `xamarin_initialize_extension_main()`.  I believe that function pointer is never being initialized, so it's trying to call through a null pointer.

And the reason I believe that pointer is never initialized is because the resulting binary contains no reference to `NSExtensionMain`.

Specifically, on the Mac, if I look in `bin/release/net6.0-ios15.0/ios-arm64/app.app/Plugins/ext.appex` the built executable is `ext`.

Running `nm -u ext | grep NSExtensionMain` yields no output.  The Mach-O executable does not reference `NSExtensionMain`.

For comparison purposes, I have an equivalent network extension in Swift, and if I go spelunking through the Xcode build output and find the corresponding executable file, in that case `nm -u` does show a reference to `NSExtensionMain`.

Furthermore, the same is true of the classic (pre-net6.0) Xamarin repro from the same zip file from #13742.
`cd bin/iPhone/Release/vpnxam.app/PlugIns/ext.appex`
`nm -u ext | grep NSExtensonMain` DOES show the reference.

And disassembling the executable with 
`otool -x -v ext > disassembly.txt`
shows the initialization of that function pointer as the first thing in the section, as perhaps one might expect, since `xamarin_initialize_extension_main` is decorated with `__attribute__ ((constructor))`

### Expected Behavior

`xamarin_extension_main` should be initialized to `NSExtensionMain`

The built executable for the extension should have a reference to `NSExtensionMain`

### Actual Behavior

No reference to `NSExtensionMain` visible. Crash. See above.

### Environment

I first reproduced this using 15.2.300-preview.14.258, installing with a rollback file from the CI builds from the fix for #13742.

In terms of inspecting the executables with `nm -u` and `otool`, the same problem can be seen with the current version obtained from `dotnet workload install ios`, 15.2.301-preview.13.7.

### Build Logs

(I can provide build logs if needed, but I'm not sure they would be helpful here?)

### Example Project (If Possible)

See the zip file I attached to #13742

### PR DESCRIPTION
**Pull request by ericsink**
Resolves #14285 

The basic changes here are to:

1.  Make sure `libextension-dotnet.a` gets built, and with the `-DEXTENSION` flag.
2.  Make sure `libextension-dotnet.a` gets included in the package alongside `libxamarin-dotnet.a`
3. At msbuild time, when `IsAppExtension` is true, include `libextension-dotnet.a` as a `NativeReference`.
4. Do all of the above for tvOS as well.

(Most of the lines changed in this PR came from @rolfbjarne -- credit to him, blame to me)

I have left come comments that I assume will be addressed and removed during review.

### CODE REVIEW

File: runtime/Makefile [hunk 1]
```diff
@@ -671,6 +693,53 @@ $(foreach platform,$(DOTNET_PLATFORMS),$(foreach rid,$(DOTNET_$(platform)_RUNTIM
...
+#
+
+define DotNetLibExtensionTemplate
+
+# TODO note that we build libextension and libtvextension
+# for all cases, even though we need at most one of them.
+# for example, the output directory for tvos contains
+# libextension-dotnet.a (which is for iOS)
+# and libtvextension-dotnet.a (which is for tvOS)
+# it seems we should only need to build one.
```
Comment (rolfbjarne) [inline runtime/Makefile:707]: We select which extension library to build for which platform here:

https://github.com/xamarin/xamarin-macios/blob/427405c53cf9f23360992bbda1bd660214f7b4ff/runtime/Makefile#L570-L573

so we should only build the ones we want (and not libtvosextension-dotnet.a for iOS for instance).

If I look for `libtvextension-dotnet*` after a build, I only get results for tvOS:

```
$ find . -name 'libtvextension-dotnet*'
./.libs/tvsimulator/libtvextension-dotnet.arm64.a
./.libs/tvsimulator/libtvextension-dotnet.x86_64.a
./.libs/tvos/libtvextension-dotnet.arm64.a
```

maybe you had leftover files from previous attempts/builds?

File: runtime/Makefile [hunk 2]
```diff
@@ -671,6 +693,53 @@ $(foreach platform,$(DOTNET_PLATFORMS),$(foreach rid,$(DOTNET_$(platform)_RUNTIM
...
+
+# note also that we build libapp-dotnet, but I'm not sure
+# we need that one at all.
+
+# note also that there is a separate set of instructions to
+# build these, above, in the monotouch section of the file,
+# the instructions here build (potentially identical) files
+# except with -dotnet included in the name.
```
Comment (rolfbjarne) [inline runtime/Makefile:715]: Yes, the logic to build the legacy Xamarin libraries will be deleted once we drop support for legacy Xamarin. Until that happens, we're going to keep both versions around.

**Review by rolfbjarne — COMMENTED** [review]
I'll have a look at the PR, and fix any issues I find.

It might take a couple of weeks though, since I'll be on vacation for Easter.

Thanks a lot for your contribution!

File: runtime/Makefile (hunk 1 above)
Comment (ericsink) [reply runtime/Makefile:707]: > maybe you had leftover files from previous attempts/builds

Possibly. I see what you mean now about how it should only be building what's needed.

**Review by rolfbjarne — APPROVED** [review]

**Review by chamons — APPROVED** [review]
LGTM

**Review by dalexsoto — APPROVED** [review]
Amazing!

**Pull request merged**